inc/namespace.php

```diff
@@ -957,21 +962,68 @@ function redact_metadata( $metadata ) {
 }
 
 /**
- * Truncate sql query to given maximum size.
+ * Maps array values using a callback recursively.
+ *
+ * @param callable $callback The function to apply to leaf nodes.
+ * @param array $array The array to modify.
+ * @return array
+ */
+function array_map_recursive( $callback, $array ) {
+	$func = function ( $item ) use ( &$func, &$callback ) {
+		return is_array( $item ) ? array_map( $func, $item ) : call_user_func( $callback, $item );
+	};
+
+	return array_map( $func, $array );
+}
+
+/**
+ * Truncate values in the metadata for a trace.
+ *
+ * @param array $trace The trace metadata.
+ * @return array
+ */
+function truncate_trace( $trace ) {
+	$max_size = 63 * 1024; // 63 KB, leaving room for UDP headers etc.
+	$trace_size = mb_strlen( json_encode( $trace ) ); // @codingStandardsIgnoreLine wp_json_encode not available.
+
+	if ( $trace_size <= $max_size ) {
+		return $trace;
+	}
+
+	// Metadata is currently the only target for truncation.
+	if ( ! isset( $trace['metadata'] ) ) {
+		return $trace;
+	}
+
+	// Find the amount of bytes we need to restrict metadata values to.
+	$metadata_size = mb_strlen( json_encode( $trace['metadata'] ) ); // @codingStandardsIgnoreLine wp_json_encode not available.
+	$metadata_max_size = $max_size - ( $trace_size - $metadata_size );
+	$metadata_count = count( $trace['metadata'], COUNT_RECURSIVE );
+	$max_value_length = floor( $metadata_max_size / $metadata_count );
+
+	$trace['metadata'] = array_map_recursive( function ( $value ) use ( $max_value_length ) {
```

Comment:
I wonder if this function can be replaced with [array_walk_recursive](https://www.php.net/manual/en/function.array-walk-recursive.php)?

Reply:
I was trying that in an iteration but I might have been doing something wrong so I switched. `array_walk_recursive` only visits leaf nodes but should be ok to do that in theory if we're passing by reference and editing in place.

Reply:
Updated 👍